### PR TITLE
make use of `RotateWithElasticsearchNodeRotation` tag more intuitive

### PR DESCRIPTION
**Pull request by twrichards**
As hinted at in the description of https://github.com/guardian/elasticsearch-node-rotation/pull/91, in this PR I aim to make the selection of which node to rotate more intuitive (**but** stopped short of eliminating the tag on the instances all together, in part because its actually used in the CFN... https://github.com/guardian/elasticsearch-node-rotation/blob/88f57c470b5775f36d5d979e537469675f5de702/cloudformation.yaml#L135 ... to restrict SSM access, but also it does provide flexibility about which specific instances are eligible for rotation).

## What does this change?
Remove `GetOldestNode` step/lambda since the behaviour was largely duplicated in the `DiscoverAutoScalingGroup` step (now renamed to `GetTargetNode`) which now...

- finds oldest node based on whether the **ASG has the** `RotateWithElasticsearchNodeRotation` tag set to "true" AND the instance has the same tag set to true - unlike the previous behaviour where it chose the ASG based on whether any instance within it had the tag set to true, then found the oldest within that ASG to actually rotate, regardless of whether that instance had the tag set to true (which is where the confusion arose)
- allows the node to be specified explicitly by `targetInstanceId` in the input object 

## How to test
Although already tested with the grid (see https://github.com/guardian/elasticsearch-node-rotation/pull/92#issuecomment-1010940122), to test simply do a partial deploy of https://riffraff.gutools.co.uk/preview/yaml?project=elasticsearch-node-rotation&build=253&stage=INFRA&updateStrategy=MostlyHarmless with your teams steps selected... then run with and without `targetInstanceId` in the input object to test the two behaviours.

## Images
image

### CODE REVIEW

File: cloudformation.yaml [hunk 1]
```diff
@@ -194,13 +194,13 @@ Resources:
...
+  GetTargetNodeLambda:
     Type: "AWS::Lambda::Function"
     DependsOn: NodeRotationLambdaRole
     Properties:
-      FunctionName: !Sub ${Stack}-enr-discover-auto-scaling-group-${Stage}
-      Description: "Looks up by tag the autoscaling groups to rotate"
-      Handler: "discoverAutoScalingGroup.handler"
+      FunctionName: !Sub ${Stack}-enr-get-target-node-${Stage}
+      Description: "Looks up by tag of the autoscaling group the most suitable instance/node to rotate"
+      Handler: "GetTargetNode.handler"
```
Comment (andrew-nowak) [inline cloudformation.yaml:203]:
```suggestion
      Handler: "getTargetNode.handler"
```

**Comment by twrichards** [comment]
✅  TESTED with grid ES cluster...

### Specific instance id as input...
image

 ### No instance id as input, so finding oldest...
image

File: cloudformation.yaml [hunk 2]
```diff
@@ -194,13 +194,13 @@ Resources:
                 - states:StartExecution
                 Resource: !Ref NodeRotationStepFunction
 
-  DiscoverAutoScalingGroupLambda:
```
Comment (jacobwinch) [inline cloudformation.yaml:197]: These CFN changes will remove/replace some lambda functions. Consequently I think we'll need to be a bit careful when rolling out this change as an in-progress execution of the Step Function might still be relying on some of the old lambdas for subsequent steps.

We could do this in two PRs:

1. Introduce all new lambdas (without deleting/replacing any old ones)
2. Delete old lambdas

Alternatively given the nature of this project, it's probably simplest to just ensure that these changes are deployed when the Step Function is not running.

File: src/addNode.ts [hunk 3]
```diff
@@ -6,19 +6,19 @@ import {ElasticsearchClusterStatus} from './elasticsearch/types';
 
 export async function handler(event: ClusterStatusResponse): Promise<AddNodeResponse> {
 
-    const oldestInstance: Instance = event.oldestElasticsearchNode.ec2Instance;
+    const targetInstance: Instance = event.targetElasticSearchNode.ec2Instance;
```
Comment (jacobwinch) [inline src/addNode.ts:9]: I prefer the `target` naming - much clearer 👍

**Comment by jacobwinch** [comment]
This looks great so far - can it be rolled out as it is, or does it need further changes?

File: cloudformation.yaml (hunk 2 above)
Comment (twrichards) [reply cloudformation.yaml:197]: yeah, given there's only a few teams using this (one of which you belong to, and I one of the others, leaving perhaps only @rtyley 's team) best just to do it at a time outside of our scheduled executions, should be easy to coordinate. that said it's not quite ready to go as I need to get round to writing up the description properly...

**Review by jacobwinch — APPROVED** [review]
Great work 💯

File: cloudformation.yaml [hunk 4]
```diff
@@ -116,7 +116,7 @@ Resources:
                 - !Sub arn:aws:autoscaling:${AWS::Region}:${AWS::AccountId}:autoScalingGroup:*:autoScalingGroupName/*
                 Condition:
                   StringLike:
-                    autoscaling:ResourceTag/RotateWithElasticsearchNodeRotation: "*"
+                    autoscaling:ResourceTag/RotateWithElasticsearchNodeRotation: "true"
```
Comment (rtyley) [inline cloudformation.yaml:119]: Just checked, we do indeed have the tag value set to `true`:

https://eu-west-1.console.aws.amazon.com/ec2autoscaling/home?region=eu-west-1#/details/Ophan-ES7-WarmDataAutoscalingGroup-1X4AC7UAK7UAN?view=details

image

**Review by rtyley — COMMENTED** [review]
Looks like it will do no harm to Ophan! Reducing the number of steps/lambdas is nice 👍 

We will deploy to Ophan and kick off a rotation now...

**Comment by rtyley** [comment]
Rotation in Ophan looks to have gone fine!

image

image

**Comment by twrichards** [comment]
> Rotation in Ophan looks to have gone fine!
> 
> image
> 
> image

Ahh that's good news, many thanks for checking @rtyley

**Pull request merged**